### PR TITLE
Improved initial metering test exception handling

### PR DESCRIPTION
**Pull request by rtamalin**
During the initial startup we perform a metering test in an effort to detect deployment failures as early as possible. If this testing encounters an error we should endeavour to update the csp_config to indicate that the error occurred before the adapter process dies.

For the sake of code readability the metering test code has been moved out of the main() routine to a new metering_test() routine.

Updated existing tests to reflect the move of the metering test to the new routine, and added an initial test to validate handling of hook.update_csp_config() failures.

Fixes: #108

### CODE REVIEW

File: tests/unit/test_adapter.py [hunk 1]
```diff
@@ -557,70 +559,159 @@ def test_main(
...
+        cba_pm.hook,
+        cba_config,
+        cba_log
+    )
+    assert cache != {}
+    assert csp_config != {}
 
-    # test catching CSP Billing Adapter exception caused by KeyError
-    with pytest.raises(SystemExit) as e:
-        cba_main()
```
Comment (KeithMnemonic) [inline tests/unit/test_adapter.py:618]: I think if you do not call main, you miss one line in the pytest run. Not sure if you care but wanted to make you aware
```
 pytest --cov=csp_billing_adapter
====================================================================== test session starts ======================================================================
platform linux -- Python 3.6.15, pytest-7.0.1, pluggy-1.0.0
rootdir: /home/kberger/repos/github/csp/csp-billing-adapter, configfile: setup.cfg, testpaths: tests
plugins: cov-4.0.0
collected 68 items

tests/unit/test_adapter.py .................                                                                                                              [ 25%]
tests/unit/test_bill_utils.py ...........                                                                                                                 [ 41%]
tests/unit/test_config.py ...                                                                                                                             [ 45%]
tests/unit/test_csp_cache.py .....                                                                                                                        [ 52%]
tests/unit/test_csp_config.py .                                                                                                                           [ 54%]
tests/unit/test_local_csp.py ....                                                                                                                         [ 60%]
tests/unit/test_memory_cache.py ....                                                                                                                      [ 66%]
tests/unit/test_memory_csp_config.py ....                                                                                                                 [ 72%]
tests/unit/test_product_api.py .                                                                                                                          [ 73%]
tests/unit/test_utils.py ..................                                                                                                               [100%]

---------- coverage: platform linux, python 3.6.15-final-0 -----------
Name                                       Stmts   Miss  Cover   Missing
------------------------------------------------------------------------
csp_billing_adapter/__init__.py                7      0   100%
csp_billing_adapter/adapter.py               148      1    99%   421
csp_billing_adapter/bill_utils.py            105      1    99%   260
csp_billing_adapter/config.py                 39      0   100%
csp_billing_adapter/csp_cache.py              30      0   100%
csp_billing_adapter/csp_config.py             11      0   100%
csp_billing_adapter/csp_hookspecs.py          13      0   100%
csp_billing_adapter/exceptions.py             12      0   100%
csp_billing_adapter/hookspecs.py              10      0   100%
csp_billing_adapter/local_csp.py              22      0   100%
csp_billing_adapter/memory_cache.py           21      0   100%
csp_billing_adapter/memory_csp_config.py      21      0   100%
csp_billing_adapter/product_api.py            13      0   100%
csp_billing_adapter/storage_hookspecs.py      16      0   100%
csp_billing_adapter/utils.py                  83      0   100%
------------------------------------------------------------------------
TOTAL                                        551      2    99%

Required test coverage of 90.0% reached. Total coverage: 99.64%

====================================================================== 68 passed in 2.15s ======================
```

Comment (rtamalin) [reply tests/unit/test_adapter.py:618]: All of the statements of main() are still called as part of the test_main() unit test case in the unit tests.

What has changed is that before, when I wanted to test the various exception cases associated with the metering testing in the main() routine,m I needed to set up the appropriate mocks for all of the stuff that main() did that I didn't want to test.

But now that the metering test has moved to a separate routine, I can test those exception cases more directly.

However, the main() routine is still tested.

The two statements in entire code base that are currently not being covered by the unit testing are:
* the call of main() from the if __name__ == '__main__' check in adapter.py
* the pass statement in the else of the if comsumption_reporting == 'volume' check in bill_utils.py.

**Review by KeithMnemonic — APPROVED** [review]
LGTM

**Review by smarlowucf — APPROVED** [review]
LGTM

**Review by brett060102 — APPROVED** [review]

**Pull request merged**